### PR TITLE
Resuming rewriting the output file contents, so previous data removed

### PR DESCRIPTION
**Pull request by xm1k3**
## Proposed changes

Solved problem related to issue #2851 

Used `os.OpenFile()` instead of `os.Create()` so now:
- if output file exist, it will append to the existing one
- if is not present in the path, it will create the new file

## Checklist

- [x] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [ ] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by Mzack9999 — APPROVED** [review]
lgtm - example:
```console
$ echo test > a.txt
$ go run . -l targets.txt -c 1 -bs 1 -o a.txt
$ go run . -l targets.txt -c 1 -bs 1 -o a.txt
$ go run . -l targets.txt -c 1 -bs 1 -o a.txt
$ cat a.txt
test
```

**Review by ehsandeep — APPROVED** [review]

**Review by ehsandeep — CHANGES_REQUESTED** [review]
@xm1k3 please ignore the above review, we need to append to file only when `-resume` option is used; similar to https://github.com/projectdiscovery/httpx/pull/823/commits/5b787384560f92a77811ad137cb95b5fb32337d8 otherwise output option defaults to write to file.

**Comment by Mzack9999** [comment]
@ehsandeep we probably should also comment on the original issue, that the wanted functionality can be obtained with:
```console
$ nuclei -json >> output.txt
```
Note: the user doesn't mention `-resume` flag

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
- [ ] merge conflict

File: v2/pkg/output/file_output_writer.go [hunk 1]
```diff
@@ -12,12 +12,20 @@ type fileWriter struct {
 }
 
 // NewFileOutputWriter creates a new buffered writer for a file
-func newFileOutputWriter(file string) (*fileWriter, error) {
-	output, err := os.Create(file)
-	if err != nil {
-		return nil, err
+func newFileOutputWriter(file string, resume string) (*fileWriter, error) {
```
Comment (Mzack9999) [inline v2/pkg/output/file_output_writer.go:15]: What if we rewrite it this way?
```go
// NewFileOutputWriter creates a new buffered writer for a file
func newFileOutputWriter(file, resume string) (*fileWriter, error) {
	var flags int
	if resume != "" {
		flags = os.O_APPEND | os.O_CREATE | os.O_WRONLY
	} else {
		flags = os.O_CREATE
	}
	output, err := os.OpenFile(file, flags, 0644)
	if err != nil {
		return nil, err
	}
	return &fileWriter{file: output}, nil
}
```

File: v2/internal/runner/runner.go [hunk 2]
```diff
@@ -186,7 +186,7 @@ func New(options *types.Options) (*Runner, error) {
 	runner.hmapInputProvider = hmapInput
 
 	// Create the output file if asked
-	outputWriter, err := output.NewStandardWriter(!options.NoColor, options.NoMeta, options.Timestamp, options.JSON, options.JSONRequests, options.MatcherStatus, options.StoreResponse, options.Output, options.TraceLogFile, options.ErrorLogFile, options.StoreResponseDir)
+	outputWriter, err := output.NewStandardWriter(!options.NoColor, options.NoMeta, options.Timestamp, options.JSON, options.JSONRequests, options.MatcherStatus, options.StoreResponse, options.Output, options.TraceLogFile, options.ErrorLogFile, options.StoreResponseDir, options.Resume)
```
Comment (Mzack9999) [inline v2/internal/runner/runner.go:189]: As we are using mostly the `options` struct, we should consider about passing it directly as argument, the function signature would be much more concise, for example:
```go
// Create the output file if asked
outputWriter, err := output.NewStandardWriter(options)
```

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
suggesting code refactoring for better readability and maintenance

File: v2/internal/runner/runner.go (hunk 2 above)
Comment (xm1k3) [reply v2/internal/runner/runner.go:189]: yes, much easier then to use, I do refactoring

File: v2/pkg/output/file_output_writer.go (hunk 1 above)
Comment (xm1k3) [reply v2/pkg/output/file_output_writer.go:15]: ok you're right, this is better, I implement it

File: v2/pkg/output/output.go [hunk 3]
```diff
@@ -123,53 +123,60 @@ type ResultEvent struct {
 }
 
 // NewStandardWriter creates a new output writer based on user configurations
-func NewStandardWriter(colors, noMetadata, timestamp, json, jsonReqResp, MatcherStatus, storeResponse bool, file, traceFile string, errorFile string, storeResponseDir string) (*StandardWriter, error) {
-	auroraColorizer := aurora.NewAurora(colors)
+//
+//	func NewStandardWriter(colors, noMetadata, noTimestamp, json, jsonReqResp, MatcherStatus, storeResponse bool, file, traceFile string, errorFile string, storeResponseDir string, resumeFlag string) (*StandardWriter, error) {
```
Comment (Mzack9999) [inline v2/pkg/output/output.go:127]: debug code left-over

**Review by Mzack9999 — CHANGES_REQUESTED** [review]

**Review by Mzack9999 — APPROVED** [review]

**Review by ehsandeep — CHANGES_REQUESTED** [review]
Output is not being written to the output file on the default run, we can include test for this.

```console
./nuclei -u example.com -o test.txt -pt ssl -silent

[ssl-dns-names] [ssl] [info] example.com [example.edu,example.com,example.org,www.example.com,www.example.edu,www.example.net,www.example.org,example.net]
[tls-version] [ssl] [info] example.com [tls13]
[deprecated-tls] [ssl] [info] example.com [tls10]
[ssl-issuer] [ssl] [info] example.com [DigiCert Inc]
[deprecated-tls] [ssl] [info] example.com [tls11]
```

```
$ wc test.txt 
0 test.txt
```

**Comment by xm1k3** [comment]
Hi @ehsandeep, i updated the code optimizing it and i also tested it, can you check again `-u example.com -o test.txt -pt ssl -silent`?

@Mzack9999 do you think it's a good optimization to handle it like this?
```go
...
...
        var output *os.File
	var err error
	if resume {
		output, err = os.OpenFile(file, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	} else {
		output, err = os.Create(file)
	}
	if err != nil {
		return nil, err
	}
...
```

**Review by Mzack9999 — APPROVED** [review]
lgtm - 
image

**Review by ehsandeep — APPROVED** [review]

**Pull request merged**